fastcrypto/src/vrf.rs

```diff
@@ -0,0 +1,268 @@
+pub trait VRFPublicKey: Serialize + DeserializeOwned + Eq + Debug {
+    type PrivateKey: VRFPrivateKey<PublicKey = Self>;
+}
+
+/// Represents a private key used to compute outputs for a verifiable random function (VRF).
+pub trait VRFPrivateKey: Serialize + DeserializeOwned + Eq + Debug {
+    type PublicKey: VRFPublicKey<PrivateKey = Self>;
+}
+
+/// A keypair for a verifiable random function (VRF).
+pub trait VRFKeyPair<const OUTPUT_SIZE: usize>: Serialize + DeserializeOwned + Eq + Debug {
+    type Proof: VRFProof<OUTPUT_SIZE, PublicKey = Self::PublicKey>;
+    type PrivateKey: VRFPrivateKey<PublicKey = Self::PublicKey>;
+    type PublicKey: VRFPublicKey<PrivateKey = Self::PrivateKey>;
+
+    /// Generate a new keypair using the given RNG.
+    fn generate<R: AllowedRng>(rng: &mut R) -> Self;
+
+    /// Generate a proof for the given input.
+    fn prove(&self, input: &[u8]) -> Self::Proof;
+
+    /// Compute both hash and proof for the given input.
+    fn output(&self, input: &[u8]) -> ([u8; OUTPUT_SIZE], Self::Proof) {
+        let proof = self.prove(input);
+        let output = proof.to_hash();
+        (output, proof)
+    }
+}
+
+/// A proof that the output of a VRF was computed correctly.
+pub trait VRFProof<const OUTPUT_SIZE: usize>: Serialize + DeserializeOwned + Eq + Debug {
+    type PublicKey: VRFPublicKey;
+
+    /// Verify the correctness of this proof.
+    fn verify(&self, input: &[u8], public_key: &Self::PublicKey) -> Result<(), FastCryptoError>;
+
+    /// Verify the correctness of this proof and VRF output.
+    fn verify_output(
+        &self,
+        input: &[u8],
+        public_key: &Self::PublicKey,
+        output: [u8; OUTPUT_SIZE],
+    ) -> Result<(), FastCryptoError> {
+        self.verify(input, public_key)?;
+        if self.to_hash() != output {
+            return Err(FastCryptoError::GeneralError);
+        }
+        Ok(())
+    }
+
+    /// Compute the output of the VRF with this proof.
+    fn to_hash(&self) -> [u8; OUTPUT_SIZE];
+}
+
+/// An implementation of an Elliptic Curve VRF (ECVRF) using the Ristretto255 group.
+/// The implementation follows the specifications in draft-irtf-cfrg-vrf-15
+/// (https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/).
+pub mod ecvrf {
+    use crate::error::FastCryptoError;
+    use crate::groups::ristretto255::{RistrettoPoint, RistrettoScalar};
+    use crate::groups::{GroupElement, Scalar};
+    use crate::hash::{HashFunction, Sha512};
+    use crate::serde_helpers::ToFromByteArray;
+    use crate::traits::AllowedRng;
+    use crate::vrf::{VRFKeyPair, VRFPrivateKey, VRFProof, VRFPublicKey};
+    use elliptic_curve::hash2curve::{ExpandMsg, Expander};
+    use serde::{Deserialize, Serialize};
+
+    /// draft-irtf-cfrg-vrf-15 specifies suites for suite-strings 0x00-0x04 and notes that future
+    /// designs should specify a different suite_string constant, so we use 0x05 here.
```

Comment:
it's not always clear or true that future designs won't skip a value, is there a reason for preferring a future (potentially never assigned value)?

Reply:
draft-irtf-cfrg-vrf-15 specifies that:
> If future designs need to specify variants (e.g., additional ciphersuites) of the RSA-FDH-VRF or the ECVRF in this document, then, to avoid the possibility that an adversary can obtain a VRF output under one variant, and then claim it was obtained under another variant, they should specify a different suite_string constant.

If I understand this correctly, we just need to choose any other suite string than the ones in the draft. But we can add a note that we update the suite_string according to new releases of the specifications.